pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go

```diff
@@ -282,6 +294,121 @@ func configMapDeleted(obj interface{}) {
+// gets added to sharedInformerFactory's Store before it's started. Then using informerManager's
+// PVCLister should find the existing PVCs as well.
+func pvcAdded(obj interface{}) {}
+
+// pvAdded adds a volume to the volumeIDToPvcMap if it's already in Bound phase.
+// This ensures that all existing PVs in the cluster are added to the map, even across container restarts.
+func pvAdded(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	pv, ok := obj.(*v1.PersistentVolume)
+	if pv == nil || !ok {
+		log.Warnf("pvAdded: unrecognized object %+v", obj)
+		return
+	}
+
+	if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == csitypes.Name &&
+		pv.Spec.ClaimRef != nil && pv.Status.Phase == v1.VolumeBound {
+
+		//Add volume handle to PVC mapping
+		objKey := pv.Spec.CSI.VolumeHandle
+		objVal := pv.Spec.ClaimRef.Namespace + "/" + pv.Spec.ClaimRef.Name
+
+		k8sOrchestratorInstance.volumeIDToPvcMap[objKey] = objVal
+		log.Debugf("pvAdded: Added '%s -> %s' pair to volumeIDToPvcMap", objKey, objVal)
+	}
+}
+
+// pvUpdated updates the volumeIDToPvcMap when a PV goes to Bound phase
+func pvUpdated(oldObj, newObj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	// Get old and new PV objects
+	oldPv, ok := oldObj.(*v1.PersistentVolume)
+	if oldPv == nil || !ok {
+		log.Warnf("PVUpdated: unrecognized old object %+v", oldObj)
+		return
+	}
+
+	newPv, ok := newObj.(*v1.PersistentVolume)
+	if newPv == nil || !ok {
+		log.Warnf("PVUpdated: unrecognized new object %+v", newObj)
+		return
+	}
+
+	// PV goes into Bound phase
+	if oldPv.Status.Phase != v1.VolumeBound && newPv.Status.Phase == v1.VolumeBound {
+		if newPv.Spec.CSI != nil && newPv.Spec.CSI.Driver == csitypes.Name && newPv.Spec.ClaimRef != nil {
+			log.Debugf("pvUpdated: PV %s went to Bound phase", newPv.Name)
+			//Add volume handle to PVC mapping
+			objKey := newPv.Spec.CSI.VolumeHandle
+			objVal := newPv.Spec.ClaimRef.Namespace + "/" + newPv.Spec.ClaimRef.Name
+
+			k8sOrchestratorInstance.volumeIDToPvcMap[objKey] = objVal
+			log.Debugf("pvUpdated: Added '%s -> %s' pair to volumeIDToPvcMap", objKey, objVal)
+		}
+	}
+}
+
+// pvDeleted deletes an entry from volumeIDToPvcMap when a PV gets deleted
+func pvDeleted(obj interface{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = logger.NewContextWithLogger(ctx)
+	log := logger.GetLogger(ctx)
+
+	pv, ok := obj.(*v1.PersistentVolume)
+	if pv == nil || !ok {
+		log.Warnf("PVDeleted: unrecognized object %+v", obj)
+		return
+	}
+	log.Debugf("PV: %s deleted. Removing entry from volumeIDToPvcMap", pv.Name)
+
+	if pv.Spec.CSI != nil {
+		delete(k8sOrchestratorInstance.volumeIDToPvcMap, pv.Spec.CSI.VolumeHandle)
```

Comment:
Where is this `delete` method defined? Shouldn't you be checking if the PV is a vSphere CSI PV?

Reply:
Can you make delete thread safe?

Reply:
`delete` is a built-in method to delete entries from map. We're checking if PV is a vSphere CSI PV while adding to the map. IMO we don't need to check it while deleting.
Added logic to make it thread safe.

Reply:
Checking for vSphere CSI PV would avoid the write lock and unlock that gets invoked as part of `remove` method.